File: go/core/cost_test.go
package core

import (
	"context"
	"math"
	"testing"
)

func TestCostTrackerAccumulates(t *testing.T) {
	var tr CostTracker
	pricing := map[string]ModelPricing{"m": {InputPerMTok: 1.0, OutputPerMTok: 2.0}}
	tr.Record("m", Usage{PromptTokens: 1_000_000, CompletionTokens: 500_000}, pricing)
	tr.Record("m", Usage{PromptTokens: 0, CompletionTokens: 500_000}, pricing)
	if tr.Usage.TotalTokens() != 2_000_000 {
		t.Fatalf("total tokens = %d, want 2,000,000", tr.Usage.TotalTokens())
	}
	if math.Abs(tr.CostUSD-3.0) > 1e-9 {
		t.Fatalf("cost = %v, want 3.0", tr.CostUSD)
	}
}

func TestCostTrackerUnknownModelNoCost(t *testing.T) {
	var tr CostTracker
	tr.Record("unknown", Usage{PromptTokens: 100, CompletionTokens: 50}, map[string]ModelPricing{})
	if tr.Usage.TotalTokens() != 150 || tr.CostUSD != 0 {
		t.Fatalf("unknown model: tokens=%d cost=%v", tr.Usage.TotalTokens(), tr.CostUSD)
	}
}

func TestCostTrackerExceeds(t *testing.T) {
	tr := CostTracker{Usage: Usage{PromptTokens: 80, CompletionTokens: 40}, CostUSD: 0.5}
	if tr.Exceeds(nil) {
		t.Fatal("nil budget should never exceed")
	}
	if tr.Exceeds(&CostBudget{MaxTokens: 200}) {
		t.Fatal("120 tokens should not exceed 200")
	}
	if !tr.Exceeds(&CostBudget{MaxTokens: 100}) {
		t.Fatal("120 tokens should exceed 100")
	}
	if tr.Exceeds(&CostBudget{MaxUSD: 1.0}) {
		t.Fatal("0.5 USD should not exceed 1.0")
	}
	if !tr.Exceeds(&CostBudget{MaxUSD: 0.5}) {
		t.Fatal("0.5 USD should exceed 0.5")
	}
}

func TestRunReportsUsageAndCost(t *testing.T) {
	client := &fakeClient{scripted: []ChatResponse{{Content: "hi", Usage: Usage{PromptTokens: 1_000_000, CompletionTokens: 1_000_000}}}}
	agent := NewSmoothAgent(client, AgentOptions{Model: "claude-haiku-4-5"})
	res, err := agent.Run(context.Background(), "hello", nil)
	if err != nil {
		t.Fatal(err)
	}
	if res.Usage.TotalTokens() != 2_000_000 {
		t.Fatalf("usage total = %d", res.Usage.TotalTokens())
	}
	// haiku default pricing = 1.0 in, 5.0 out per 1M.
	if math.Abs(res.CostUSD-6.0) > 1e-9 {
		t.Fatalf("cost = %v, want 6.0", res.CostUSD)
	}
	if res.BudgetExceeded {
		t.Fatal("budget should not be exceeded")
	}
}

func TestRunStopsWhenBudgetExceeded(t *testing.T) {
	client := &fakeClient{scripted: []ChatResponse{
		{ToolCalls: []ToolCall{{ID: "c1", Name: "noop", Arguments: "{}"}}, Usage: Usage{PromptTokens: 200}},
	}}
	agent := NewSmoothAgent(client, AgentOptions{Model: "claude-haiku-4-5", Budget: &CostBudget{MaxTokens: 100}})
	res, err := agent.Run(context.Background(), "go", nil)
	if err != nil {
		t.Fatal(err)
	}
	if !res.BudgetExceeded || res.Iterations != 1 || res.ToolCalls != 0 {
		t.Fatalf("expected budget stop at iter 1 with 0 tool calls; got %+v", res)
	}
}

File: typescript/core/test/cost.test.ts
import { describe, expect, it } from 'vitest';
import { AgentOptions, ChatClientLike, SmoothAgent } from '../src/agent.js';
import { CostBudget, CostTracker, totalTokens, Usage } from '../src/cost.js';

describe('CostTracker', () => {
    it('accumulates usage and cost', () => {
        const t = new CostTracker();
        const pricing = { m: { inputPerMTok: 1.0, outputPerMTok: 2.0 } };
        t.record('m', { promptTokens: 1_000_000, completionTokens: 500_000 }, pricing);
        t.record('m', { promptTokens: 0, completionTokens: 500_000 }, pricing);
        expect(totalTokens(t.usage)).toBe(2_000_000);
        expect(t.costUsd).toBeCloseTo(3.0); // 1*1 + 2*1
    });

    it('counts tokens for an unknown model but costs nothing', () => {
        const t = new CostTracker();
        t.record('unknown', { promptTokens: 100, completionTokens: 50 }, {});
        expect(totalTokens(t.usage)).toBe(150);
        expect(t.costUsd).toBe(0);
    });

    it('budget exceed logic', () => {
        const t = new CostTracker();
        t.usage = { promptTokens: 80, completionTokens: 40 };
        t.costUsd = 0.5;
        expect(t.exceeds(undefined)).toBe(false);
        expect(t.exceeds({ maxTokens: 200 })).toBe(false);
        expect(t.exceeds({ maxTokens: 100 })).toBe(true);
        expect(t.exceeds({ maxUsd: 1.0 })).toBe(false);
        expect(t.exceeds({ maxUsd: 0.5 })).toBe(true);
    });
});

// fake client carrying usage
function fake(scripted: Array<{ content: string | null; tool_calls?: unknown; usage?: { prompt_tokens: number; completion_tokens: number } }>): ChatClientLike {
    let i = 0;
    return {
        chat: {
            completions: {
                create: async () => {
                    const s = scripted[i++];
                    return { choices: [{ message: { content: s.content, tool_calls: s.tool_calls as never } }], usage: s.usage };
                },
            },
        },
    };
}

describe('SmoothAgent cost integration', () => {
    it('reports usage and cost on a run', async () => {
        const client = fake([{ content: 'hi', usage: { prompt_tokens: 1_000_000, completion_tokens: 1_000_000 } }]);
        const agent = new SmoothAgent(client, { model: 'claude-haiku-4-5' } satisfies AgentOptions);
        const res = await agent.run('hello');
        expect(totalTokens(res.usage)).toBe(2_000_000);
        expect(res.costUsd).toBeCloseTo(1.0 + 5.0); // haiku default (1 in, 5 out)
        expect(res.budgetExceeded).toBe(false);
    });

    it('stops when the budget is exceeded', async () => {
        const client = fake([
            { content: null, tool_calls: [{ id: 'c1', function: { name: 'noop', arguments: '{}' } }], usage: { prompt_tokens: 200, completion_tokens: 0 } },
        ]);
        const budget: CostBudget = { maxTokens: 100 };
        const agent = new SmoothAgent(client, { model: 'claude-haiku-4-5', budget });
        const res = await agent.run('go');
        expect(res.budgetExceeded).toBe(true);
        expect(res.iterations).toBe(1);
        expect(res.toolCalls).toBe(0);
    });
});
